libraries/app/database_api.cpp

```diff
@@ -1772,108 +1794,100 @@ vector<blinded_balance_object> database_api_impl::get_blinded_balances( const fl
+   if( queue.size() )
    {
-      map< pair<asset_id_type, asset_id_type>, vector<variant> > broadcast_queue;
-      for( const auto& obj : objs )
-      {
-         const limit_order_object* order = dynamic_cast<const limit_order_object*>(obj);
-         if( order )
-         {
-            auto sub = _market_subscriptions.find( order->get_market() );
+      auto capture_this = shared_from_this();
+      fc::async([capture_this, this, queue](){
+          for( const auto& item : queue )
+          {
+            auto sub = _market_subscriptions.find(item.first);
             if( sub != _market_subscriptions.end() )
-               broadcast_queue[order->get_market()].emplace_back( order->id );
-         }
-      }
-      if( broadcast_queue.size() )
-      {
-         auto capture_this = shared_from_this();
-         fc::async([capture_this,this,broadcast_queue](){
-             for( const auto& item : broadcast_queue )
-             {
-               auto sub = _market_subscriptions.find(item.first);
-               if( sub != _market_subscriptions.end() )
-                   sub->second( fc::variant(item.second ) );
-             }
-         });
-      }
+                sub->second( fc::variant(item.second ) );
+          }
+      });
    }
 }
 
+void database_api_impl::on_objects_removed( const vector<object_id_type>& ids, const vector<const object*>& objs )
+{
+   handle_object_changed(_notify_remove_create, ids, [objs](object_id_type id) -> const object* {
+
+      auto it = std::find_if(
+            objs.begin(), objs.end(), 
+            [id](const object* o) {return o != nullptr && o->id == id;});
+
+      if (it != objs.end())
+         return *it;
+
+      return nullptr;
+   });
+}
+
+void database_api_impl::on_objects_new(const vector<object_id_type>& ids)
+{
+   handle_object_changed(_notify_remove_create, ids, 
+      std::bind(&object_database::find_object, &_db, std::placeholders::_1)
+   );
+}
+
 void database_api_impl::on_objects_changed(const vector<object_id_type>& ids)
 {
-   vector<variant>    updates;
-   map< pair<asset_id_type, asset_id_type>,  vector<variant> > market_broadcast_queue;
+   handle_object_changed(false, ids, 
+      std::bind(&object_database::find_object, &_db, std::placeholders::_1)
+   );
+}
 
-   for(auto id : ids)
+void database_api_impl::handle_object_changed(bool force_notify, const vector<object_id_type>& ids, std::function<const object*(object_id_type id)> find_object)
+{
+   if( _subscribe_callback )
    {
-      const object* obj = nullptr;
-      if( _subscribe_callback )
+      vector<variant> updates;
+      
+      for(auto id : ids)
       {
-         obj = _db.find_object( id );
-         if( obj )
+         const object* obj = nullptr;
+         if( force_notify || is_subscribed_to_item(id) )
          {
-            updates.emplace_back( obj->to_variant() );
-         }
-         else
-         {
-            updates.emplace_back(id); // send just the id to indicate removal
```

Comment:
This is important, removals should only broadcast the id of the object, not the full object.